_posts/2020-10-02-producer-tuning.md

```diff
@@ -0,0 +1,307 @@
+
+Compressing data batches using the `compression.type` property improves throughput and reduces the load on storage, but might not be suitable for low-latency applications where the cost of compression or decompression is prohibitive.
+
+Message compression adds latency in the producer (CPU time spent compressing the messages),
+but makes requests (and potentially disk writes) smaller, which can increase throughput.
+
+Use the `compression.type` property to specify a valid compression codec.
+You can choose `gzip`, `snappy`, `lz4`, or `zstd`, each of which have varying compression speeds.
+
+```properties
+compression.type=gzip
+batch.size=32000
+```
+
+If you think compression is worthwhile, the best type of compression to use will depend on the messages being sent.
+
+> **Adding threads**
+>
+> Compression is handled on the thread calling `KafkaProducer.send()`, so if the latency of this method matters for your application you can add more more threads.
+
+#### Pipelining messages
+
+Pipelining might sound like it has something to do with surfing the famous _Pipeline_ reef in Hawaii,
+but it's actually the equally impressive method of sending more requests from producers before the response to a previous request has been received.
+
+For pipelining we use our old friend the `max.in.flight.requests.per.connection` property.
+You might recall its contribution to ordered delivery earlier in this post.
+
+Moving message requests along more frequently will obviously improve throughput. But there is a point at which you might see less beneficial effects, such as less efficient batching.
+
+#### Adjusting message delivery waiting time
+
+Improve throughput of your message requests by adjusting the maximum time to wait before a message is delivered and completes a send request.
+
+Use the `delivery.timeout.ms` property to specify the maximum time in milliseconds to wait for a complete send request.
+You can set the value to `MAX_LONG` to delegate to Kafka an indefinite number of retries.
+
+You can also direct messages to a specified partition by writing a custom partitioner to replace Kafka's default, and specify the class name using the `partitioner.class` property. A custom partitioner allows you to choose how you map messages to partitions, based on the data in the message.
+
+```properties
+# ...
+delivery.timeout.ms=120000
+partitioner.class=my-custom-partitioner
+# ...
+```
+
+### Adapt to survive
+
+Fine tuning your producers helps alleviate performance issues.
+But don't be tempted to make a few adjustments and think your work is done.
+
+You should consider fine tuning as part of a continual optimization process. Monitor Kafka regularly. Look for changing trends in usage and investigate how fine tuning can help your Kafka deployment adapt. But, as you'll know, this is only one half of the story.
+
+Next time we'll look at how you can optimize your consumers.
+
+Be sure to check back for the next installment.
```

Comment:
Should we promise this?

Reply:
I think so. I have a _to do_ to expand the section in the guide to include consumers. Might need some help though.